### PR TITLE
PLAT-430: Add conveyor limiter to control max parallel processing

### PR DESCRIPTION
**Pull request by MikhailZhukov**
Add hierarchical semaphore for limit runner parallelisms

### CODE REVIEW

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type ParallelProcessingLimiter interface {
+	SyncLink() smachine.SyncLink
+	NewChild(childValue int, name string) smsync.SemaChildLink
+}
+
+func NewParallelProcessingLimiter(maxTasks int) ParallelProcessingLimiter {
+	if maxTasks <= 0 {
+		maxTasks = runtime.NumCPU() - 1
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:22]: -2

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	}
+	if maxTasks < 1 {
+		maxTasks = 1
+	}
+	return &conveyorSemaphore{
+		semaphore: smsync.NewSemaphoreWithFlags(maxTasks, "ConveyorParallelProcessingLimiter", smsync.QueueAllowsPriority),
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:28]: 1. factually, this is a limiter for runner, not for conveyor
2. make the name an exported constant
3. add a helper method to receives this link pls

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+type conveyorSemaphore struct {
+	semaphore smsync.SemaphoreLink
+}
+
+func (cs conveyorSemaphore) SyncLink() smachine.SyncLink {
+	return cs.semaphore.SyncLink()
+}
+
+func (cs conveyorSemaphore) NewChild(childValue int, name string) smsync.SemaChildLink {
+	return cs.semaphore.NewChildExt(true, childValue, name, smsync.AllowPartialRelease)
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:43]: + PrioritizePartialAcquire

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	}
+}
+
+var _ ParallelProcessingLimiter = conveyorSemaphore{}
+
+type conveyorSemaphore struct {
+	semaphore smsync.SemaphoreLink
+}
+
+func (cs conveyorSemaphore) SyncLink() smachine.SyncLink {
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:38]: nope, no one should be able to directly have this syncLink
do helper methods instead pls

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+package conveyor
+
+import (
+	"runtime"
+
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine/smsync"
+)
+
+type ParallelProcessingLimiter interface {
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:15]: do you really need an interface? 
as you can't mock SemaChildLink anyway, this is futile

File: ledger-core/conveyor/conveyor.go [hunk 6]
```diff
@@ -55,6 +55,7 @@ type PulseConveyorConfig struct {
 	SlotMachineConfig                 smachine.SlotMachineConfig
 	EventlessSleep                    time.Duration
 	MinCachePulseAge, MaxPastPulseAge uint32
+	MaxParallelTasks                  int
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor.go:58]: IT MUST NOT BE HERE

File: ledger-core/conveyor/conveyor_semaphore.go (hunk 1 above)
Comment (MikhailZhukov) [reply ledger-core/conveyor/conveyor_semaphore.go]: done

File: ledger-core/conveyor/conveyor_semaphore.go (hunk 2 above)
Comment (MikhailZhukov) [reply ledger-core/conveyor/conveyor_semaphore.go]: done

File: ledger-core/conveyor/conveyor_semaphore.go (hunk 3 above)
Comment (MikhailZhukov) [reply ledger-core/conveyor/conveyor_semaphore.go]: done

File: ledger-core/conveyor/conveyor_semaphore.go (hunk 4 above)
Comment (MikhailZhukov) [reply ledger-core/conveyor/conveyor_semaphore.go]: done

File: ledger-core/conveyor/conveyor_semaphore.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+func (cs ParallelProcessingLimiter) AcquireForThisStepAndRelease(ctx smachine.SynchronizationContext) smachine.BoolDecision {
+	return ctx.AcquireForThisStepAndRelease(cs.semaphore.SyncLink())
+}
+
+func (cs ParallelProcessingLimiter) Release(ctx smachine.SynchronizationContext) bool {
+	return ctx.Release(cs.semaphore.SyncLink())
+}
+
+func (cs ParallelProcessingLimiter) NewChild(childValue int, name string) smsync.SemaChildLink {
```
Comment (cyraxred) [inline ledger-core/conveyor/conveyor_semaphore.go:53]: may be NewHierarchySemaphore for clarity?

**Pull request merged**